src/util.rs

```diff
@@ -16,8 +22,69 @@ pub(crate) fn load_rustdoc_from_file(path: &Path) -> anyhow::Result<Crate> {
         .read_to_string(&mut s)
         .with_context(|| format!("Failed to read rustdoc JSON output file {:?}", path))?;
 
-    serde_json::from_str(&s)
-        .with_context(|| format!("Failed to parse rustdoc JSON output file {:?}", path))
+    match serde_json::from_str(&s) {
+        Ok(value) => Ok(value),
+        Err(_) => {
+            // Attempt to figure out the more precise reason the deserialization failed.
+            // Several possible options and their resolutions:
+            // (1) The file isn't actually a rustdoc JSON file. The user should supply a valid file.
+            // (2) The rustdoc JSON file has a version number that is too old, and isn't supported.
+            //     The user should upgrade to a newer nightly Rust version and regenerate the file.
+            // (3) The rustdoc JSON file has a version number that is too new, and isn't supported.
+            //     The user should attempt to upgrade to a newer cargo-semver-checks version
+            //     if one is already available, or open a GitHub issue otherwise.
+
+            // The error on this line is case (1).
+            let version = serde_json::from_str::<RustdocFormatVersion>(&s).with_context(|| {
+                format!(
+                    "Failed to parse the rustdoc JSON file, and could not determine its \
+                    format version. Are you sure this is a valid rustdoc JSON file? \
+                    File path: {:?}",
+                    path
+                )
+            })?;
+
+            match version.format_version.cmp(&rustdoc_types::FORMAT_VERSION) {
+                std::cmp::Ordering::Less => {
+                    // The error here is case (2).
+                    bail!(
+                        "Failed to parse rustdoc JSON file: {path:?}\n\n\
+                        This version of cargo-semver-checks requires rustdoc JSON version {0}, \
+                        but the JSON file uses the older version {1}.\n\n\
+                        Please upgrade to a Rust nightly version where rustdoc outputs \
+                        JSON format version {0}.\n\n\
+                        It's usually easiest to upgrade to the latest versions of \
+                        both cargo-semver-checks and Rust nightly, \
+                        which should always be compatible with each other.",
+                        rustdoc_types::FORMAT_VERSION,
+                        version.format_version,
+                    )
+                }
+                std::cmp::Ordering::Greater => {
+                    // The error here is case (3).
+                    bail!(
+                        "Failed to parse rustdoc JSON file: {path:?}\n\n\
+                        This version of cargo-semver-checks requires rustdoc JSON version {0}, \
+                        but the JSON file uses the newer version {1}.\n\n\
```

Comment:
I feel like these messages could be more concise so the user can more easily find the information they need.

Something like
```
rustdoc output is too new (version {}) for cargo-semver-checks when parsing {}

note: a newer version of cargo-semver-checks is likely available
```
- cargo's style guide says to not capitalize error messages
- rustc's style guide says to avoid "did you mean" or "suggestions" but to instead offer people alternatrives (ie be less prescriptive) which is why I made it a "note:".  cargo doesn't follow this atm but I suspect we should conform to it.

Reply:
I tried my hand at mirroring your style, let me know if this is better.